File: misc-problems/cartesian-product/cartesianProduct.js
function cartesianProduct(arr1, arr2) {
    let product = [];

    for (let num1 of arr1) {
        for (let num2 of arr2) {
            product.push([num1, num2]);
        }
    }

    return product;
}


// Alternate solution
function cartesianProduct2(arr1, arr2) {
    return arr1.reduce((acc, curr) => {
        return acc.concat(arr2.map(num => [curr, num]));
    }, []);
}

File: misc-problems/cartesian-product/cartesianProduct.py
def cartesian_product(*args):
    if not args:
        return
    if len(args) == 1:
        for i in args[0]:
            yield (i,)
    else:
        for i in args[0]:
            for j in cartesian_product(*args[1:]):
                yield (i,) + j
 

if __name__ == '__main__':
    for i in cartesian_product([1, 2, 3], [4, 5], [6, 7]):
        print(i)